designs/node-overlays.md

```diff
@@ -0,0 +1,186 @@
+
+### Misconfiguration
+
+Given our merge semantics, it's possible for customers to define multiple NodeOverlays with undefined behavior. In both cases, we will identify and communicate these misconfigurations, but should we fail open or fail closed? Failing closed means that a misconfigured would halt Karpenter from launching new Nodes until the misconfiguration is resolved. Failing open means that Karpenter might launch capacity that differs from what an operator expects, and result in thrashing scale in/outs.
+
+We choose to fail closed to prevent undefined behavior that could incur costs. However, we have the option to fail closed for the entire provisioning loop, or for a single NodePool. For example, if multiple NodeOverlays match NodePoolA with the same weight, but only a single NodeOverlay matches NodePoolB, Karpenter could exclude NodePoolA as invalid, but proceed with NodePoolB for simulation. These semantics are very similar to how Karpenter deals with `.spec.limits` or `InsufficientCapacityErrors`. See: https://karpenter.sh/docs/concepts/scheduling/#fallback.
+
+### Observability
+
+How can users be aware of how their NodeOverlays have applied to their NodePools? Following from our decision above, we will model the application of NodeOverlays at the NodePool level. This leads us to observe the behavior of NodeOverlays in the status of the NodePool.
+
+For example:
+```
+kind: NodePool
+status:
+  conditions:
+    - type: InstanceTypesReady
+      status: True
+  nodeOverlays:
+    - default
+    - extended-resources
+  instanceTypes:
+    - m5.large
+	- ...
+```
+
+Exposing the full details of the mutations caused by NodeOverlays cannot be feasibly stored in ETCD. Further, we do not want to require that customers read the Karpenter logs to debug this feature. When GetInstanceTypes is called per NodePool, we will emit an event on the NodePool to surface deeper semantic information about the effects of the NodeOverlays. Our implementation will need to be conservative on the volume and frequency of information, but provide sufficient insight for customers to verify that their configurations are working as expected.
+
+### Per-Field Semantics
+
+The semantics of each field carry individual nuance. Price could be modeled directly via a `price` field, but we can significantly reduce the number of NodeOverlays that need to be defined if we instead model the semantic as a function of the currently defined price (e.g. `pricePercent` or `priceAdjustment`). Similarly, `capacity` and `overhead` make sense to be modeled directly, but `capacityPercent` makes much less sense than `overheadPercent`.
```

Comment:
Is it possible to lay out what you are thinking the API looks like for each field? I am not sure I grok what we will support. 
Are we supporting something like:
```
overhead:
  - memory: 10%
```

Second I think the names can be confusing at the moment, maybe its just me.

For example, if I look just a pricePercent does that mean percent of discount or percent of price? 
For overhead its a similar story, is this overhead the amount I am adding to a resource or overhead that I need to subtract from the resource itself.

Lastly, if I specify a capacity that exist within the scheduling simulation for an instance type but its incorrect, such as huge pages, do I override that value or can I even do that? If not can I specify a negative overhead to add value to an existing resource?

Reply:
Got it. I realize now I totally missed that section right under.

I think I understand the pricePercent a bit better now so I am happy with keeping that name. Other names I was thinking about was `priceAdjustmentPercent` which indicates you are adjusting the price based on input % rather than a percent of price. Happy to either way